sql-plugin/src/main/330db/scala/org/apache/spark/sql/rapids/execution/GpuExecutorBroadcastHelper.scala

```diff
@@ -24,14 +24,29 @@ import org.apache.spark.sql.catalyst.expressions.Attribute
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
+/**
+ * Helper object to handle [[EXECUTOR_BROADCAST]] from Databricks. Logically, an executor broadcast 
+ * involves the executor "broadcasting" its data to the other executors, and then each of the 
+ * executors build their own version of the broadcast relation. We accomplish a similar version 
+ * of this by reading the shuffle data in each executor and passing the columnar batch data to the
+ * GPU versions of BroadcastHashJoin and BroadcastNestedLoopJoin, which by necessity must build 
+ * these relations on the executor because they do the work of the join on the GPU device itself,
+ * which means they require a format of the data that can be used on the GPU.
+ */
 object GpuExecutorBroadcastHelper extends Arm {
 
+  private def shuffleDataIterator(shuffleData: RDD[ColumnarBatch]): Iterator[ColumnarBatch] = {
+    shuffleData.partitions.map { part =>
+      shuffleData.iterator(part, TaskContext.get())
```

Comment:
GPU columnar batches are not directly serializable.  Does this have issues when running in a distributed setup?  Local mode won't expose that problem since it often avoids going through serialization, as it's all in a single JVM.

Reply:
I've not seen any issues with a distributed setup (and I did use this code in #7533, which I tested in distributed cluster with 4 workers, with and without the RAPIDS multithreaded shuffle as well). It's probably working because it's only used in executor broadcast scenario (ie it's code that's only running on the executor that's calling it). And in executor broadcast (from what I can tell from Databricks) I think all of the executors read shuffle data directly and bypass using the cached RDD, and use `getShuffleRDD`. It's somewhat redundant and it's not quite the coordinated executor-side broadcast described in presentations on the subject, but I'm pretty sure that's what's going on.

But I should clean up some of the code using `getShuffleRDD` (which is what Databricks uses) instead of `doExecute`. I think that will do better in terms of the Coalesce

Reply:
How is that possible? ColumnVector is not Serializable.

https://github.com/rapidsai/cudf/blob/20c945be7295efbd6c55d4f388bfcf898484915a/java/src/main/java/ai/rapids/cudf/ColumnVector.java#L41
https://github.com/rapidsai/cudf/blob/20c945be7295efbd6c55d4f388bfcf898484915a/java/src/main/java/ai/rapids/cudf/ColumnView.java#L31

So unless CUDF is some how serializing things that are not Serializable this is going to fail.  Even if Spark did that the only thing that it is holding onto is a pointer to memory in the given process it came from. When you went to deference the pointer you would get garbage, and hopefully crash.

If I let this through on #7533 that was a serious bug and we need to fix it there too. I have no idea how this could ever work.

Reply:
OK I thought about this more. And it might not be as bad as we think. I see two possibilities for how this might work.

The RDD that is being read includes the shuffle and enough other processing to turn it into the RDD[ColumnarBatch] that we want. If so, then it will just read the shuffle the data. That should be fine.

The other possibility is that the RDD does not include the shuffle at all. In that case because the RDD is lazy I think it might compute the RDD inline to produce the iterator. In that case if we had to do a join to compute the data, then the join would be computed on each task that is trying to read in the data.

If this is working, then I really hope that it is the first option, but we should check out the construction of the RDD to see exactly what it is.